### ISSUE TITLE
Bug: `reportUnusedDisableDirectives: "error"` should cause "warn" rules to run even in `--quiet`

### ISSUE THREAD

**Issue by robyoder** [eslint/eslint#19625]
### Environment

Node version: v20.18.0
npm version: v10.8.2
Local ESLint version: v9.24.0 (Currently used)
Global ESLint version: Not found
Operating System: darwin 24.3.0

### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration</summary>

```js
[
	{
		linterOptions: {
			reportUnusedDisableDirectives: "error",
		},
		rules: {
			"no-debugger": [ "warn" ]
		}
	},
]
```
</details>

```js
// eslint-disable-next-line no-debugger
console.log("not a debugger statement");
```

Run `npx eslint --quiet`

### What did you expect to happen?

When running `npx eslint --quiet` on this code, I would expect to receive an error about an unused disable directive because `reportUnusedDisableDirectives` is set to `error`:

```
error  Unused eslint-disable directive (no problems were reported from 'no-debugger')
```

### What actually happened?

I only receive the error when running without `--quiet`.

### Link to Minimal Reproducible Example

https://gist.github.com/robyoder/13b651b65ab3e46bb68f738678042774

### Participation

- [x] I am willing to submit a pull request for this issue.

### Additional comments

This appears to have been caused by [a change in ESLint 9](https://github.com/eslint/eslint/pull/17274) to not run rules set to `"warn"` when using `--quiet`. Because that could result in false positives for `reportUnusedDisableDirectives`, those warn-level rules are simply ignored, resulting in false negatives instead.

I would suggest that if `reportUnusedDisableDirectives` is set to `"error"`, rules set to `"warn"` should be run even in quiet mode so that unused directives can still be detected.

**Comment by Tanujkanti4441** [comment]
Hi @robyoder, thanks for the issue.

This works as intended, as you can see in [this](https://github.com/eslint/eslint/issues/16450#issuecomment-1286780419) discussion. Since the `--quiet` flag disables rules with a severity of `warn`, only rules configured as error will be reported. So, it was decided to ignore the `eslint-disable` for the rules that are set to `warn` so they won't get flagged by `reportUnusedDisableDirectives`.

**Comment by Tanujkanti4441** [comment]
Okay I got it. The intention was to ignore the `eslint-disable` that are useful, but now it's also ignoring the `unused` ones. It's indeed a bug for me.

**Comment by mdjermanovic** [comment]
This works as intended, per the design from the RFC:: https://github.com/eslint/rfcs/blob/d88b1948b37d1e28669145f0534d1dd824a0e1cc/designs/2023-only-run-reporting-rules/README.md?plain=1#L65-L71

The documentation for the [`--quiet` option](https://eslint.org/docs/latest/use/command-line-interface#--quiet) states that only rules set to error will be run, so we have to assume that additional "meta" checks that need the rules to be run, like the check for unused disable directives, would pass for rules set to warn.

> I would suggest that if `reportUnusedDisableDirectives` is set to `"error"`, rules set to `"warn"` should be run even in quiet mode so that unused directives can still be detected.

While this reasoning definitely makes sense, as not running rules indirectly suppresses meta checks that could result in error-level lint messages, I'm in favor of the current behavior for performance reasons, and, as stated in the paragraph from the RFC linked above, "the user likely does not expect `warn` flags to be touched at all in this mode".

Curious what others think about this.

**Comment by robyoder** [comment]
I understand the intention. My point is that it is a violation of user expectations to see errors reported without `--quiet` but not with `--quiet`. It's a similar surprise in an IDE where both warnings and errors are shown. Even if it is not the default, it should be _possible_ to configure ESLint to report only errors but run all rules so that the unused meta-rule is accurate.

**Comment by snitin315** [comment]
I’m also in favor of the existing behavior. When rules like `no-debugger` are skipped under `--quiet`, we can't accurately assess whether disables like `// eslint-disable-next-line no-debugger` are actually needed. Running warning-severity rules just to catch unused disables when `reportUnusedDisableDirectives` is set to `error` would also introduce performance trade-offs.

**Comment by nzakas** [comment]
I agree with @mdjermanovic and @snitin315. This is a pretty extreme edge case and while it may violate expectations of some users, I think this is ultimately the correct approach given the tradeoffs.
